src/ads/RewardedAd.ts

```diff
@@ -133,7 +136,10 @@ export class RewardedAd extends MobileAd implements MobileAdInterface {
     );
   }
 
-  onAdEvent(handler: AdEventListener) {
+  /**
+   * @deprecated Use addAdEventListener instead.
+   */
+  onAdEvent(handler: AdEventHandler) {
```

Comment:
This makes a breaking change (because of the argument type) at the same time as a deprecation if I read correctly?
Why not make the new type a new name vs changing the type name here, then it's just a deprecation mark

Additionally, this could be a facade over the new implementation couldn't it? That is, you swap the current implementation here for an implementation that iterated over the enum of types calling `addAdEventListener` for each type and returned a function that called all of the resulting unsubscribe functions. If I understand correctly that would simulate the old behavior? Could add a little bit of doc that says calling it clears all existing users to make it bug-compatible, and do that first even - clear map, iterate over event types adding listener for all types, return method that wraps all the unsubscribes so caller calls that and all listeners are removed ?

What do you think? I typically prefer deprecating an API and re-implementing it underneath with the new methods to give them the workout they need and to avoid maintaining multiple "real" implementations vs one new/"real" and one wrapper with deprecated facade

Reply:
> This makes a breaking change (because of the argument type)

Since the argument type (which was AdEventListener before) has not been exposed to external users before(the type AdEventListener is not exported in index.ts), I think it is not breaking change, just a internal type name change.

> If I understand correctly that would simulate the old behavior?

Yes, it simulates the old behavior which returns unsubscribe function for the listener. But, the unsubscribe function does not unsubscribe all the registered listeners, but unsubscribes only for a listener registered with the method. As you can see in https://github.com/invertase/react-native-google-mobile-ads/blob/5bb68a1bc1fc67f480c8ad8638f7a0b9129367f2/src/ads/MobileAd.ts#L130-L135 To unsubscribe all the listeners, callers have to save reference for the all of the unsubscribe functions that is returned from each registration of listener, then iterate over the functions to unsubscribe each listeners.

If it is needed, we can implement additional method which can unsubscribe all of the registered listeners.

Reply:
Ah - okay re: the type change, that works for me, thank you for the explanation - sorry I did not notice that

Let me start from first principles on the general idea of going with the new implementation (which I like!) and re-implementing the deprecated methods via emulation:

- old add listener method: *replaced* all subscribers, and listener is responsible for figuring out types - it is called for all events
- function returned from old add listener method: unsubscribes from events, and since an add replaces existing listeners this means all listeners are unsubscribed

Do I understand that correctly? If not nothing else I type will be valid - but assuming I understand it correctly, then emulation of old style with new methods is:

- deprecated add listener: removes all existing listeners, iterates over all types, subscribes to all, calls provided listener with type + event from any of the underlying events
- function returned from deprecated listener: removes all listeners for all types

I think that maps the new style to the old style and can be done ? It should have a note in the deprecation describing exactly what it does and how callers should migrate to the new per-event / multiple-listener style

> we can implement additional method which can unsubscribe all of the registered listeners.

I think this would be an ergonomic / developer experience benefit, and could be used in any "shim" that adapted the old style to the new style internally as part of the deprecation

Alternatively: could just do a breaking change and purge the old API methods immediately with a very clear prescription in a v5->v6 migration blurb that showed old code -> new code and then we don't worry about emulation or multiple implementations or anything. I'm okay with breaking change as a general rule - especially if all the old features exist - as long as there is *clear* helpful guidance for consumers.

Reply:
Re: breaking change or emulating old behavior:

> Which is the best way do you think?

This is tough for me. I understand it is developer-unfriendly to do breaking changes but also, this is to add new features and I just released a 5.1.1 that should be stable for people to stay on for a while.

I think with a really good migration guide maybe we just break. Make a docs v5->v6 and in there include a clear description of the motivation for the change, *exactly* how the behavior changes, and a quick snippet of old code vs new code (including correct unsubscription), update the example app (and mention people can look at the example app diff for a migration example) and then we just move on.

I'm sensitive to library consumer pain but I also know that our audience right now is not so big, so I am prioritizing maintainer effort a bit over consumer pain just now

Reply:
> our audience right now is not so big

I agree to this;
I'll work on purge the `onAdEvent`, add method to remove all the listeners, and make migration docs.